[PERF] stock: reduce memory usage of `_get_orderpoint_action`

pivi-odoo opened the pull request:
## Issue
On databases with lots of products (400k+ `product.product`), the action Inventory > Operations > Replenishment takes a lots of memory to complete, leading to "Out of Memory" errors in some use cases.

## Analysis
When triggering the action, we need to recompute the `virtual_available` of a lot of products (regular occurrence when a DB for example has a lot of variants). The compute `_compute_quantities` uses the ORM `filtered` method to filter out `services` out of the `product.product` we have. This method is memory hungry, as for each `_prefetch_ids` (capped at 1000, cf. `PREFETCH_MAX`) it will store in cache the `product_tmpl_id.type`, which stores all fields of the product_template + the type, bloating the cache with *useless* data, as it's not used beyond this filtering process.

## Solution
Disable the prefetcher, the bulk of the work is done in `_compute_quantities_dict`, which makes a lot of `read_group` and doesn't benefit much from fields already cached (this function itself also disables the prefetcher internally at some point). Also correcting the looping order in `_get_orderpoint_action` to avoid multiple cache invalidation when those can be batched in 1 call. The `PREFETCH_MAX` makes sure the compute is only executed with 1k records at a time, so explicitly dividing `all_products_ids` into batches of 5k is unnecessary.

## Results
- Memory usage of `_get_orderpoint_action`:

| Memory usage | Before  | After   |
|--------------|---------|---------|
| Total Memory | 1.8 GiB | 363 MiB |
| Allocations  | 10170   | 2804     |

(profiler data available on referenced ticket)

- Execution time of `_get_orderpoint_action`:

| `product.product` count | Before   | After    |
|-------------------------|----------|----------|
| 1k                      | 1.3s     | 1.1s     |
| 10k                     | 4.1s     | 3.8s     |
| 100k                    | 30s      | 25.9s    |
| 400k+                   | 2min 25s | 1min 39s |

As we can observe, we've reduced drastically the memory usage of the function and also the number of allocations quite drastically. As a side effect there is also a slight speed improvement at small scale, and at large scale the gains are up to ~40% faster.

## Reference
opw-3415087

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/135807).

Whenrow commented:
this basically revert https://github.com/odoo/odoo/commit/e382b70c699d8a0ef0abc8f2a96c904b8abce9a2 that has been made to avoid MemoryError on the same amount of product (400k). 
@ryv-odoo  can we have your opinion on this ?  I don't want to reintroduce previous bad behaviour. Thanks

pivi-odoo commented:
Thanks @ryv-odoo and @Whenrow for your input.

I've taken Remy's comment and incorporated them into the patch.

I've moved the `prefetch_fields=False` to the problematic filter so it doesn't leak the context later in execution and re-enable the prefetcher post filtering.

The invalidation of the recordset was not needed indeed, since I was made aware that context dependent computes are recomputed without the need of explicit invalidation.

I've run new benchmarks (uploaded on ticket), there is no major regression in term of memory usage. Just a bit more allocations, but the total memory usage remained the same. The commit msg was adjusted with the new values. In term of time there is no change.

I will leave up to the R&D if they want to adapt with the `fetch` API in master :)

Whenrow commented:
robodoo r+
